src/server/gc_worker/applied_lock_collector.rs

```diff
@@ -0,0 +1,732 @@
+    ) {
+        if cf != CF_LOCK {
+            return;
+        }
+
+        let max_ts = self.load_max_ts();
+        if max_ts.is_zero() {
+            return;
+        }
+
+        let locks: Result<Vec<_>> = kv_pairs
+            .iter()
+            .map(|(key, value)| {
+                Lock::parse(value)
+                    .map(|lock| (key, lock))
+                    .map_err(|e| ErrorInner::Mvcc(e.into()).into())
+            })
+            .filter(|result| result.is_err() || result.as_ref().unwrap().1.ts <= max_ts)
+            .map(|result| {
+                // `pre_apply_plain_keys` will be invoked with the data_key in RocksDB layer. So we
+                // need to remove the `z` prefix.
+                result.map(|(key, lock)| (Key::from_encoded_slice(origin_key(key)), lock))
+            })
+            .collect();
+
+        match locks {
+            Err(e) => self.send(LockObserverMsg::Err(e)),
+            Ok(l) => self.send(LockObserverMsg::Locks(l)),
+        }
+    }
+
+    fn pre_apply_sst(&self, _: &mut ObserverContext<'_>, cf: CfName, _path: &str) {
+        if cf == CF_LOCK {
+            let e = box_err!("snapshot of lock cf applied from sst file");
+            error!("cannot collect all applied lock"; "err" => ?e);
+            self.send(LockObserverMsg::Err(e));
+        }
+    }
+}
+
+struct LockCollectorRunner {
+    observer_state: Arc<LockObserverState>,
+
+    collected_locks: Vec<(Key, Lock)>,
+
+    /// `is_clean` is true, only it's sure that all applying of stale locks (locks with start_ts <=
+    /// specified max_ts) are monitored and collected. If there are too many stale locks or any
+    /// error happens, `is_clean` must be set to `false`.
+    /// TODO: when dirty, stop collecting.
+    is_clean: bool,
+}
+
+impl LockCollectorRunner {
+    pub fn new(observer_state: Arc<LockObserverState>) -> Self {
+        Self {
+            observer_state,
+            collected_locks: vec![],
+            is_clean: true,
+        }
+    }
+
+    fn handle_observer_msg(&mut self, msg: LockObserverMsg) {
+        if !self.is_clean {
+            return;
+        }
+
+        match msg {
+            LockObserverMsg::Err(e) => {
+                self.is_clean = false;
+                info!("lock collector marked dirty because received error"; "err" => ?e);
+            }
+            LockObserverMsg::Locks(mut locks) => {
+                if locks.len() + self.collected_locks.len() > MAX_COLLECT_SIZE {
+                    self.is_clean = false;
+                    info!("lock collector marked dirty because received too many locks");
+                    locks.truncate(MAX_COLLECT_SIZE - self.collected_locks.len());
+                }
+                self.collected_locks.extend(locks);
+            }
+        }
+    }
+
+    fn start_collecting(&mut self, max_ts: TimeStamp) -> Result<()> {
+        info!("start collecting locks"; "max_ts" => max_ts);
+        self.collected_locks.clear();
```

Comment:
What if there is a repeated start request comming?

Reply:
Oh, I think I need to check whether max_ts equals to the current max_ts. In this case we shouldn't restart.

Reply:
I think it's only safe to start it when current max_ts is 0. But if we reject all start requests when current max_ts is not 0, then we cannot start it again if we failed to stop it before... I don't have a clear mind of a good solution.